clang/lib/Sema/SemaTemplateInstantiate.cpp

```diff
@@ -4077,6 +4084,7 @@ Sema::SubstExpr(Expr *E, const MultiLevelTemplateArgumentList &TemplateArgs) {
   TemplateInstantiator Instantiator(*this, TemplateArgs,
                                     SourceLocation(),
                                     DeclarationName());
+  Instantiator.setEvaluateConstraints(false);
```

Comment:
I believe this ends up being incorrect?  We end up calling SubstExpr pretty frequently 'inside' of other constraint expression checking I think, and I think we set whether to evaluate constraints with further checks based on the current instantiator.

So I think this ends up failing to check constraints in a cascading way.  I DO know I tried something similar ot this patch at one point when doing the deferred diagnostics and ran into trouble.

IF we are going to do something like this (as I don't have sufficient evidence here to block I think), we need to hold this off until after the release so it has time to bake.

Reply:
I will sanity-check some call sites and try to break it with tests to understand this better.
The reason why I went with this change, is because of this comment on the `SubstConstraintExpr`:
```cpp
  // Unlike the above, this evaluates constraints, which should only happen at
  // 'constraint checking' time.
```
The 'above' mentioned here is `SubstExpr`. If we don't update this function, I think we should update the comment (and remove `SubstConstraintExpr` as both functions are exactly the same right now).

Reply:
Yeah, that comment is a little out of date I think.  I initially split those two with the intent of making them do something different (and infact, almost exactly the change you made!), but ended up with a ton of test failures, particularly in libc++ for a while.

I think i ended up picking up most/all of those tests as lit tests during my implementation, but I'm not confident in it enough to have us merge this before the release branch next month.

Reply:
Oh, and another signal that gave me optimism is that all existing tests pass. However, I am not sure if they are extensive.

Reply:
Oh, absolutely!  I'm reassured somewhat by  that fact of course, I was as comprehensive as I could be at the time.  BUT when doing this the first time, I ended up with about a dozen reverts trying to work through it, so there is definitely possibility that I missed some.  I'm reasonably confident/ok with this patch, just not until the release branch is taken.

Reply:
I have spent some time to understand this better and I agree that changing the default behavior of `SubstExpr` is quite a bet. So I made a more localized fix instead, please take a look. I believe we don't need to wait for a branch cut to land this change.

I have also changed `SubstConstraintExpr` to simply call `SubstExpr` and updated the comment to avoid confusion (as mentioned before, the comment used to say they're different, but they are actually the same).

I have also added tests to make sure the uses of `requires-expr` in SFINAE contexts decltype does not cause issues, I believe it tests other code paths a bit.